File: LeetCode/LeetCode/Stack/MinStack.swift
//
//  MinStack.swift
//  LeetCode
//
//  Created by Pranshu Goyal on 10/04/20.
//  Copyright © 2020 Pranshu Goyal. All rights reserved.
//

class MinStack {

	private var array: [Int] = []
	private var min = Int.max

    func push(_ x: Int) {
		guard !array.isEmpty else {
			min = x
			array.append(x)
			return
		}

		if x < min {
			array.append(2*x - min)
			min = x
		} else {
			array.append(x)
		}
    }

    func pop() -> Int {
		let y = array.popLast()!
		if y < min {
			let t = min
			min = 2*min - y
			return t
		} else {
			return y
		}
    }

    func top() -> Int {
		let y = array.last!
		if y < min {
			return min
		} else {
			return y
		}
    }

    func getMin() -> Int {
		return min
    }
}

File: python/MinStack.py

#
# MinStack.py
#

class MinStack:
	
	def __init__(self):
		self.list = []
		self.minStack = []

	def push(self, v):
		self.list.append(v)
		if len(self.minStack) == 0 or v <= self.minStack[-1]:
			self.minStack.append(v)

	def pop(self):
		v = self.list.pop()
		if v <= self.minStack[-1]:
			self.minStack.pop()
		return v

	def getMin(self):
		return self.minStack[-1]

if __name__ == "__main__":
	test = [3,2,1,4,5,2,8]
	stack = MinStack()
	for i in test:
		stack.push(i)

	print(stack.pop())
	print(stack.pop())
	print(stack.pop())
	print("Min: ", stack.getMin())
	print(stack.pop())
	print("Min: ", stack.getMin())
	print(stack.pop())
	print("Min: ", stack.getMin())
	print(stack.pop())
	print("Min: ", stack.getMin())
	print(stack.pop())
